File: design-pattern/bridge-pattern/bridge.cpp
#include <iostream>
#include <string>

using namespace std;

class Engine
{
public:
    virtual void startEngine() = 0;
};

class PetrolEngine : public Engine
{
public:
    void startEngine()
    {
        cout << "petrol engine is starting" << endl;
    }
};

class DisealEngine : public Engine
{
public:
    void startEngine()
    {
        cout << "Diseal Engine is starting" << endl;
    };
};

class ElectricEngine : public Engine
{
public:
    void startEngine()
    {
        cout << "motor is rotating" << endl;
    };
};

class Vehicle
{
protected:
    Engine *engine;

public:
    Vehicle(Engine *engine) : engine(engine) {};

    virtual void startCar() = 0;
};

class Sedan : public Vehicle
{
public:
    Sedan(Engine *engine) : Vehicle(engine) {};

    void startCar() override
    {
        engine->startEngine();
    }
};

class SUV : public Vehicle
{
public:
    SUV(Engine *engine) : Vehicle(engine) {};

    void startCar() override
    {
        engine->startEngine();
    }
};

int main(void)
{
    Engine *petrolEngine = new PetrolEngine();
    Vehicle *bmw = new SUV(petrolEngine);
    bmw->startCar();
    return 0;
}

File: design-pattern/bridge-pattern/bridge.ts
abstract class Engine {
    abstract StartEngine(): void;
};

class PetrolEngine extends Engine {
    StartEngine(): void {
        console.log("starting the petrol engine")
    };
};

class DisealEngine extends Engine {
    StartEngine(): void {
        console.log("starting the Diseal Engine");
    }
};

class ElectricMotor extends Engine {
    StartEngine(): void {
        console.log("Starting the Electric motor");
    }
};


abstract class Vechicle {
    protected engine: Engine;

    constructor(engine: Engine) {
        this.engine = engine;
    }

    abstract StartVechicle(): void;
};

class Sedan extends Vechicle {

    constructor(engine: Engine) {
        super(engine);
    };

    StartVechicle(): void {
        this.engine.StartEngine();
    }
};

class Suv extends Vechicle {
    constructor(engine: Engine) {
        super(engine);
    }

    StartVechicle(): void {
        this.engine.StartEngine();
    }
}


const engine = new PetrolEngine();
const suv = new Suv(engine);
suv.StartVechicle();